kubeflow/core/util.libsonnet

```diff
@@ -31,5 +41,11 @@
       else [],
   }.result,
 
-
+  local e = self,
+  local k = import "k.libsonnet",
+  list:: std.prune(k.core.v1.list.new([
```

Comment:
Could we make this and then in the .libsonnet files just define a list property that is the result of calling this function?

Would that lead to infinite recursion?

Reply:
I wonder if we'd be better off just manually defining the list as opposed to auto-constructing it from all the keys in the dict?

Its possible in the future we'd want to have more than 1 prototype defined in a .libsonnet file composed of different subsets of elements defined in the .libsonnet file. In which case we wouldn't want to add all the iems in the dictionary.

Reply:
There is an example of this in spartacus.libsonnet of a 'prototype' (see AllResources::) and i'm also converting tensorboard.libsonnet to use these patterns. Tensorboard is a great use case since it has {params, gcpParams, s3params} and {parts, s3parts, gcpParts}. But to answer your question:
> I wonder if we'd be better off just manually defining the list as opposed to auto-constructing it from all the keys in the dict?

For organization by 'prototype' (or grouping) you basically do:
```
{
  PrototypeA:: if params.emitPrototypeA then (self + {
    ConfigmapA: {
    },
    ServiceA: {
    },
  }),
  PrototypeB:: if params.emitPrototypeB then (self + {
    ConfigmapB: {
    },
    ServiceB: {
    },
  }),
}
```

This still allows auto-construction of lists and may be better than how tensorboard is currently grouping prototypes (which is quite clever). I think a side-by-side comparison of this pattern versus tensorboard's current approach will be helpful. 

Your other question:
> Could we make this and then in the .libsonnet files just define a list property that is the result of calling this function? Would that lead to infinite recursion?
 
All lower case fields (leading character) including hidden functions are ignored in util.list - similar to golang where exported fields have a leading upper character. It's also difficult to access util's list unless you do it explicitly (eg something like List:: util.list) since list is not added to self until 'new' returns. 

I understand why you are suggesting explicit lists instead of auto-construction (opaqueness, risk of infinite recursion). I'm hoping that we can eliminate current "bad" practices sneaking into components that involve adding params, env vars, etc (which you've suggested is a goal of this work). This is why i'm using certain conventions like upper/lower in lieu of being explicit. This auto-construction may do no more than obfuscate manifest generation. I think tensorboard will be a good use case to examine to see if these patterns (including auto-construction) are making things easier, harder or just plain confusing.

Would it be ok to wait for my update of tensorboard.libsonnet?